common/src/main/java/org/conscrypt/ConscryptEngine.java

```diff
@@ -1372,14 +1372,17 @@ public void onSSLStateChange(int type, int val) {
 
     @Override
     public boolean onNewSessionEstablished(long sslSessionNativePtr) {
+        SslSessionWrapper sessionWrapper;
         try {
-            // Cache the newly established session.
-            AbstractSessionContext ctx = sessionContext();
-            ctx.cacheSession(SslSessionWrapper.newInstance(sslSessionNativePtr, sslSession));
-            return true;
-        } catch (Exception ignored) {
+            sessionWrapper = SslSessionWrapper.newInstance(sslSessionNativePtr, sslSession);
+        } catch (Throwable e) {
```

Comment:
Hrm. This is still kinda screwy. If `newInstance` throws in `activeSession.getSessionContext()` (sorry, I missed that there was something before the constructor call), then this works. But if it throws in `Impl()`, at `SSL_SESSION_get_version`, the finalizer will still run, right? Which means that `Impl` has taken ownership of the pointer and we double-free.

I'm not sure I see the point of making the return value match the C one. You don't match the `out_copy` thing and it's the same deal. Again, the return value *doesn't mean success or failure*. It's purely a weird lifetimes thing. In a language like Java, it doesn't make sense to handle lifetimes by tossing booleans around, so it doesn't make sense to keep the return value.